Sources/LeafKit/LeafTag.swift

```diff
@@ -0,0 +1,40 @@
+public protocol LeafTag {
+    func render(_ ctx: LeafContext) throws -> LeafData
+}
+
+public var defaultTags: [String: LeafTag] = [
+    "lowercased": Lowercased(),
+    "index": LoopIndex(),
+    "isFirst": LoopIsFirst(),
+    "isLast": LoopIsLast()
+]
+
+struct Lowercased: LeafTag {
+    func render(_ ctx: LeafContext) throws -> LeafData {
+        guard let str = ctx.parameters.first?.string else {
+            throw "unable to lowercase unexpected data"
+        }
+        return .init(.string(str.lowercased()))
+    }
+}
+
+struct LoopIndex: LeafTag {
+    func render(_ ctx: LeafContext) throws -> LeafData {
+        guard let index = ctx.data["index"] else { throw "Loop accessor called on non-loop" }
```

Comment:
Throwing raw strings as errors is a rather unfortunate choice 😕. I would hope for something like this (sketched out in comment field, may have no relation whatsoever to valid code):
```swift
public struct LeafError: Error {
    public init(_ message: String, templateLocation: LeafSourceLocation) { ... }
    public let templateFile: String // the template being lexed/parsed/rendered
    public let sourceLine: UInt // the physical line of the template responsible for the error
    public let sourceColumn: Uint // the nearest physical column in the physical line that can be easily found as the source of the error, maybe 0 if it's hard to figure out
}
```
For this to work properly Leaf would have to do something Swift's compiler (and most others) do - maintain "source location" data on every single individual AST node of any kind (an expression has a source location at the point where the expression starts, each term of the expression has a location where that term appears, and so on). Import nodes and such would change the "current" file within the context of parsing (for the purposes of recording source locations), usually you'd maintain a stack.

Reply:
(Hopefully it's obvious that I think Leaf absolutely should do this, despite the extra infrastructure needed 🙂.)

Reply:
Agreed - see PR #46 for an initial start to adding a LeafError dealio.

Reply:
(Resolving as I consider that outside the scope of this PR)